liquidprompt

```diff
@@ -3835,6 +3849,113 @@ _lp_temperature_color() {
     lp_temperature_color="${LP_MARK_TEMP}${ret}${lp_temperature}°${NO_COL}"
 }
 
+######################
+# Current Disk Usage #
+######################
+
+# Convert bytes in human-readable, binary-prefixed notation (using powers of 2).
+# Precision can be 0, 1 or 2 digits. If not given, precision is two digits.
+__lp_bytes_to_human() { # bytes, [precision] -> lp_bytes, lp_bytes_units
+    local i="$1" prec="${2:-2}" size remainder=""
+    # This has to be split on multiple lines for Zsh 5.0 (see wiki)
+    local units=()
+    # Not much point in going past PiB, as Bash and Zsh start failing int math.
+    units=("B" "KiB" "MiB" "GiB" "TiB" "PiB" "EiB" "YiB" "ZiB")
+
+    for ((size=0; i > 1024 && size < ${#units[@]}-1; size++)); do
+        remainder=$((i % 1024 * 100 / 1024))
+        i=$((i / 1024))
+    done
+
+    if [[ -n $remainder ]]; then
+        if ((remainder < 10)); then
+            remainder="0$remainder"
+        fi
+    else
+        # Bytes case.
+        remainder="00"
+    fi
+
+    # Precision formating.
+    remainder="${remainder:0:$prec}"
+    if [[ -n $remainder ]]; then
+        remainder=".$remainder"
+    fi
+
+    lp_bytes="$i$remainder"
+    lp_bytes_units="${units[$size]}"
```

Comment:
Needs the array offset to be correct.

```suggestion
    lp_bytes_units="${units[$size+$_LP_FIRST_INDEX]}"
```

A good question is why the tests do not catch this issue. The answer is that the tests set the Zsh option `ksh_arrays`, which makes Zsh act like Ksh/Bash in terms of array offsets. This is to make writing tests a lot easier. 

The correct solution is to add a diff like
```diff
diff --git a/liquidprompt b/liquidprompt
--- a/liquidprompt
+++ b/liquidprompt
@@ -88,7 +88,11 @@ elif test -n "${ZSH_VERSION-}" ; then

     _LP_MARK_SYMBOL='%%'

-    _LP_FIRST_INDEX=1
+    if [[ -o ksh_arrays ]]; then
+      _LP_FIRST_INDEX=0
+    else
+      _LP_FIRST_INDEX=1
+    fi
     _LP_PERCENT='%%'
     _LP_BACKSLASH='\'
```
To the setup block of `liquidprompt`, and move the `. liquidprompt` line in each test file to after when `setopt shwordsplit ksh_arrays` is run. Which I will do after merging this PR.

That won't start catching these issues though, but at least it will make Liquid Prompt work with people running Zsh with `ksh_arrays` set, and will make tests not fail. I'm open to ideas of how to start catching this, but my only idea is to add `+$_LP_FIRST_INDEX` to each array usage in test files and stop setting `ksh_arrays`, but that is annoying.

Reply:
Can't we setup some shell check to spot array indexing without the `+` sign?

Reply:
Hm, yeah likely. Maybe a simple regex would work?

Ok, I couldn't make a simple one, but this works (ish):
```bash
$ grep '${[[:alnum:]_]\+\[[^+]\+\]}' liquidprompt | grep -v '${[[:alnum:]_]\+\[@]}' | grep -v '${[[:alnum:]_]\+\[\*]}'

            key="${fields[i]}"
                    filled+="${fillchars_as_text[i,i]}"
    lp_bytes_units="${units[$size]}"
        command=${history[$HISTCMD]}
            if [[ "${options[promptpercent]}" == on ]]; then
                local value="${precmd_functions[i]}"
                local value="${preexec_functions[i]}"
                    local value="${PROMPT_COMMAND[i]}"
```

Unfortunately, all of those (except for the line that this comment thread is about) are false positives; either because they are specific to one shell, or the `_LP_FIRST_INDEX` was already accounted for (eg `filled+="${fillchars_as_text[i,i]}"`).

A real solution would be to add a new Shellcheck rule somehow, which would allow us to add ignore directives for that rule. Unfortunately Shellcheck does not yet support custom rules, so this may just be something we need to check manually.

Reply:
This custom regexp with an inline comment for disabling it?
```sh
filled+="${fillchars_as_text[i,i]}" # _LP_FIRST_INDEX accounted.
```

Reply:
Yeah, that would be a good solution, then we can simply check for the line containing `_LP_FIRST_INDEX` anywhere in it.
```bash
grep '${[a-zA-Z0-9_]\+\[[^]@\*]\+\]}' liquidprompt | grep -v _LP_FIRST_INDEX
```

Reply:
I propose we merge this PR as is (with the failing tests) and work on this on another one. Should I rebase/squash, now?